grass.tools: Migration guide from run_command (#6862)
OSGeo/grass: doc/migration_to_grass_tools_python_api.md

avoiding repetition of the parameters.
On the other hand, if you need to set the special parameter for only one tool call,
you will typically need to create a separate *Tools* object for that call.
However, the objects are cheap to create,
so the overhead is negligible in most cases.

### Checking return codes without exceptions

By default, *grass.tools* raises exceptions when tools fail,
which is appropriate for most workflows.
If you need to check return codes without raising exceptions,
configure the *Tools* object accordingly.

The *run_command* approach uses *errors="status"*:

```python
returncode = gs.run_command(
    "r.mask.status", 
    flags="t", 
    errors="status"
)
# Returns: 1 (if no mask exists)
```

Configure the *Tools* object to ignore errors and always return result objects:

```python
tools_with_returncode = Tools(
    errors="ignore", 
    consistent_return_value=True, 
)
result = tools_with_returncode.r_mask_status(flags="t")
returncode = result.returncode
# Returns: 1
```

While the new tool call above involves more objects,
namely a separate *Tools* object and also tool result object,

> an actual tool call in the code may be quite straigforward:

Reviewer: ```suggestion
an actual tool call in the code may be quite straightforward:
```

# Returns: 1
```

### Using the overwrite flag

The overwrite flag can be set per tool call or globally for all calls.

The *run_command* approach requires it on every call:

```python
gs.run_command(
    "r.random.surface",
    output="surface",
    seed=42,
    overwrite=True,
)
```

Or you can set that for the whole process using an environment variable,
called `GRASS_OVERWRITE`.
You also can pass a dedicated environment with the *env* parameter
with the variable set there influencing only those tools for
which you pass the *env* variable.

The environment variable approach is still available with *grass.tools*,
but the *Tools* class allows a more clear and friendlier approach.

You can still pass it per call:

```python
tools = Tools()
tools.r_random_surface(output="surface", seed=42, overwrite=True)
```

However, you can set overwrite globally when creating the *Tools* object:

```python
tools = Tools(overwrite=True)
tools.r_random_surface(output="surface", seed=42)
```

This approach is particularly useful in scripts where you frequently recreate outputs
during development and testing,
but are not ready to turn your script into a GRASS tool in Python

> which provide the overwrite settings automatically.

Reviewer: ```suggestion
which provides the overwrite settings automatically.
```

```

While the *grass.tools* return value is supposed to be predictable and useful,
it is not as consistent as with the *run_command* family of functions.
If you need a consistent return value, e.g., for automation,
a consistent return value behavior can be requested:

```python
tools = Tools(consistent_return_value=True)
result = tools.r_slope_aspect(elevation="elevation", slope="slope")
# result is now a result object, even though stdout was empty
```

This configurable behavior avoids cluttering interactive sessions when a tool
creates output files, but prints nothing to stdout (like *r.slope.aspect*),
while at the same time it accommodates more advanced cases
when consistent, not just predictable, return value is needed.
This further provides better control when NumPy arrays
are returned by the tool call.

## Complete migration example

Here's a complete example showing the transformation
from *run_command* to *grass.tools* syntax for a script
which explicitly sets up and passes a session to tool calls.
(A code for an analytical GRASS tool will typically leave out the session
handling completely relying on the parent session.

> Scripts may only setup the a global session without passing it explicitly.

Reviewer: Please check if there's a missing word. Otherwise, maybe the following fixes it
```suggestion
Scripts may only set up a global session without passing it explicitly.
```
